Convert SymInts to SymFloats with SymPy

isuruf opened the pull request:
Fixes #109365


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

isuruf committed:
Convert SymInts to SymFloats with SymPy

lezcano reviewed (commented):
It is a bit hacky, but it's better than nothing? Let's wait see if @ezyang is fine with this fix.

Otherwise, we can have a more targeted fix by replacing `b == 1.0` by `b == 1` (same with the other conditions) in the reference for `pow`.

ezyang reviewed (approved):
Yeah sure, why not.

asmeurer commented:
> Otherwise, we can have a more targeted fix by replacing b == 1.0 by b == 1 (same with the other conditions) in the reference for pow.

Not sure what code this is referring to, but this reminds me that the next version of SymPy is going to change it so that `Float(1.0) == Integer(1)` gives False. I'll try to check if SymPy master breaks any of the torch tests.

isuruf commented:
Hmm, that seems to be at odds with `Eq(Symbol("x", integer=True), 1.0)` returning None instead of False.

asmeurer commented:
> Hmm, that seems to be at odds with Eq(Symbol("x", integer=True), 1.0) returning None instead of False.

There's a *lot* of background discussion for this change. The change was made here https://github.com/sympy/sympy/pull/25614 but you can see the several issues that are cross-referenced (the main one being https://github.com/sympy/sympy/issues/11707). 

The main point is

- `==` in SymPy means structural, not mathematical equality. Just as `x*(y + z) != x*y + x*z` because they are mathematically the same, but we want to be able to represent them distinctly, we want `Integer(1)` and `Float(1.0)`. On the other hand, `Eq` is symbolic mathematical equality (which is why it is usually unevaluated). 
- `None` in the SymPy assumptions system means "could be True or False". So there's no contradiction with making `Float(1.0).is_integer = None`. We did consider making it `True` instead (https://github.com/sympy/sympy/pull/25865), we rejected that change for now at least because it has more implications. 

With that being said, this hasn't been released yet so we're not sure how it will affect downstream code. I do want to at least check if it affects pytorch.

isuruf commented:
> == in SymPy means structural, not mathematical equality. Just as x*(y + z) != x*y + x*z because they are mathematically the same, but we want to be able to represent them distinctly, we want Integer(1) and Float(1.0). On the other hand, Eq is symbolic mathematical equality (which is why it is usually unevaluated).

No objections from me. We had this in symengine from the start.

> None in the SymPy assumptions system means "could be True or False".

Okay, I didn't realize that `Eq(Integer(1), Float(1.0))` will still return True in the next release.

asmeurer commented:
> Okay, I didn't realize that Eq(Integer(1), Float(1.0)) will still return True in the next release.

Yes it does, because Eq(a, b) checks `(a - b).is_zero`, and `1 - 1.0` still returns `0.0` (and `Float(0.0).is_zero` does still give `True`). In fact 0.0 is the only float that is treated as rational=True in the assumptions system, which is why I wanted to change it for all floats to be consistency. But for now, we are conservatively keeping it this way and using None for the rest.

asmeurer commented:
>  I'll try to check if SymPy master breaks any of the torch tests.

I'm having some difficulties running the torch tests (quite a few tests fail in main with symPy 1.12, most likely due to the hardware I'm testing on), but so far I did find that `test/inductor/test_torchinductor_dynamic_shapes.py::TestInductorDynamicCPU::test_shape_as_constant_reciprocal_float_exp_cpu` fails in SymPy master because of this line
https://github.com/pytorch/pytorch/blob/e360f4c6dd3f0c6a7b949a26cc3bf4fa03a6dd1b/torch/_inductor/codegen/common.py#L305

lezcano commented:
@asmeurer please submit fixes for the issues you find

lezcano commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but it is failing in trunk https://hud.pytorch.org/pytorch/pytorch/commit/0ec66b3be5a53ab960872981b5027c49c2e6b7e9, probably a landrace as this is not failing on your PR' -c landrace

jon-chuang commented:
This seems to be causing the failure: 

`test/test_proxy_tensor.py::TestProxyTensorOpInfoCPU::test_make_fx_symbolic_exhaustive_std_cpu_float32 - RuntimeError: prims::var() Expected a value of type 'float' for argument 'correction' but instead found type 'SymFloat'.`

lezcano commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3044,10 +3044,13 @@ def replace(expr, repl):
         out = bound_sympy(new_expr, new_range_env)
-        _assert_bound_is_rational(new_expr, out)
+        try:
+            _assert_bound_is_rational(new_expr, out)
```
Nope. Add the smallest possible exception within `assert_bound_is_rational`, please.

isuruf replied:
Unlike the sqrt exception added already, I don't think it makes sense to add exceptions here because the expression can be anything. I added a commit that might be a better compromise.

ezyang replied:
Compromise seems ok to me

isuruf committed:
dont expect rational in guard_float

ezyang commented:
@pytorchbot merge

ezyang commented:
failure may be legit, haven't seen it before

isuruf committed:
Skip signal_windows_bartlett test with torchinductor

This test used to fail with torchinductor because of the assertion
failure that looks for a rational. With this being fixed, now the
test is actually run with torchinductor and fails. Since
signal_windows_exponential is also skipped, we skip
signal_windows_bartlett too.

isuruf commented:
I skipped that test. The test used to fail with torchinductor because of the assertion failure that looks for a rational. With this being fixed, now the test is actually run with torchinductor and fails. Since signal_windows_exponential is also skipped, I skipped
signal_windows_bartlett too. Let me know if that's okay.

ezyang commented:
@pytorchbot merge
